### PR TITLE
resources: convert responses to TypeScript

### PR DESCRIPTION
**Pull request by quisquous**

### CODE REVIEW

**Comment by quisquous** [comment]
ugh git

```diff
diff --git a/resources/responses.js b/resources/responses.ts
index 42c73b0f3..332d635fa 100644
--- a/resources/responses.js
+++ b/resources/responses.ts
@@ -18,8 +18,18 @@
 // function that sets outputStrings and returns an object without doing
 // anything with data or matches.  See `responses_test.js`.
 
+import { LocaleText, ResponseOutput, ResponseFunc, TriggerFunc, TargetedMatches, Output } from '../types/trigger';
+import { Data } from '../types/data';
 import Outputs from './outputs';
 
+type TargetedResponseOutput = ResponseOutput<TargetedMatches>;
+type TargetedResponseFunc = ResponseFunc<TargetedMatches>;
+type TargetedFunc = TriggerFunc<TargetedMatches>;
+type StaticResponseFunc = ResponseFunc<unknown>;
+
+type Severity = 'info' | 'alert' | 'alarm';
+type SevText = 'infoText' | 'alertText' | 'alarmText';
+
 export const builtInResponseStr = 'cactbot-builtin-response';
 
 // All valid trigger fields.
@@ -58,120 +68,130 @@ export const triggerOutputFunctions = [
   'sound',
 ];
 
-export const severityMap = {
+export const severityMap: { [sev in Severity]: SevText } = {
   'info': 'infoText',
   'alert': 'alertText',
   'alarm': 'alarmText',
 };
 
-const getText = (sev) => {
+const getText = (sev: Severity): SevText => {
   if (!(sev in severityMap))
     throw new Error(`Invalid severity: ${sev}.`);
   return severityMap[sev];
 };
 
-const defaultInfoText = (sev) => {
+const defaultInfoText = (sev: Severity): SevText => {
   if (!sev)
     return 'infoText';
   return getText(sev);
 };
 
-const defaultAlertText = (sev) => {
+const defaultAlertText = (sev: Severity): SevText => {
   if (!sev)
     return 'alertText';
   return getText(sev);
 };
 
-const defaultAlarmText = (sev) => {
+const defaultAlarmText = (sev: Severity): SevText => {
   if (!sev)
     return 'alarmText';
   return getText(sev);
 };
 
-const getTarget = (matches) => {
+const getTarget = (matches: TargetedMatches) => {
   // Often tankbusters can be casted by the boss on the boss.
   // Consider this as "not having a target".
   if (!matches || matches.target === matches.source)
-    return null;
-  return matches.target || matches[1];
+    return;
+  // TODO: remember to land this matches[1] removal separately.
+  return matches.target;
 };
 
-const getSource = (matches) => {
-  return matches.source || matches[0];
+const getSource = (matches: TargetedMatches) => {
+  return matches?.source;
 };
 
 // FIXME: make this work for any number of pairs of params
-const combineFuncs = function(text1, func1, text2, func2) {
-  const obj = {};
+const combineFuncs = function(text1: SevText, func1: TargetedFunc,
+    text2: SevText, func2: TargetedFunc) {
+  const obj: TargetedResponseOutput = {};
 
   if (text1 !== text2) {
     obj[text1] = func1;
     obj[text2] = func2;
   } else {
-    obj[text1] = (data, matches, output) => {
+    obj[text1] = (data: Data, matches: TargetedMatches, output: Output) => {
       func1(data, matches, output) || func2(data, matches, output);
     };
   }
   return obj;
 };
 
-const isPlayerId = (id) => {
-  return id[0] !== '4';
+const isPlayerId = (id?: string) => {
+  return id && id[0] !== '4';
 };
 
 // For responses that unconditionally return static text.
-const staticResponse = (field, text) => (data, _, output) => {
-  // cactbot-builtin-response
-  output.responseOutputStrings = {
-    text: text,
-  };
-  return {
-    [field]: (data, _, output) => output.text(),
+const staticResponse = (field: SevText, text: LocaleText): StaticResponseFunc => {
+  return (data: unknown, _: unknown, output: Output) => {
+    // cactbot-builtin-response
+    output.responseOutputStrings = {
+      text: text,
+    };
+    return {
+      [field]: (data: unknown, _: unknown, output: Output) => output.text?.(),
+    };
   };
 };
 
-export const Responses = {
-  tankBuster: (targetSev, otherSev) => {
+type SingleSevToResponseFunc = (sev: Severity) => TargetedResponseFunc | StaticResponseFunc;
+type DoubleSevToResponseFunc = (targetSev: Severity, otherSev: Severity) => TargetedResponseFunc;
+type ResponsesMap = {
+  [response: string]: SingleSevToResponseFunc | DoubleSevToResponseFunc;
+};
+
+export const Responses: ResponsesMap = <const>{
+  tankBuster: (targetSev: Severity, otherSev: Severity) => {
     const outputStrings = {
       noTarget: Outputs.tankBuster,
       busterOnYou: Outputs.tankBusterOnYou,
       busterOnTarget: Outputs.tankBusterOnPlayer,
     };
 
-    const targetFunc = (data, matches, output) => {
+    const targetFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (!target) {
         if (data.role !== 'tank' && data.role !== 'healer')
           return;
-        return output.noTarget();
+        return output.noTarget?.();
       }
 
       if (target === data.me)
-        return output.busterOnYou();
+        return output.busterOnYou?.();
     };
 
-    const otherFunc = (data, matches, output) => {
+    const otherFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (!target) {
         if (data.role === 'tank' || data.role === 'healer')
           return;
-        return output.noTarget();
+        return output.noTarget?.();
       }
       if (target === data.me)
         return;
 
-      return output.busterOnTarget({ player: data.ShortName(target) });
+      return output.busterOnTarget?.({ player: data.ShortName(target) });
     };
 
     const combined = combineFuncs(defaultAlertText(targetSev), targetFunc,
         defaultInfoText(otherSev), otherFunc);
-    return (data, _, output) => {
+    return (data: unknown, _: unknown, output: Output): TargetedResponseOutput => {
       // cactbot-builtin-response
       output.responseOutputStrings = outputStrings;
       return combined;
     };
   },
-  tankBusterSwap: (busterSev, swapSev) => {
+  tankBusterSwap: (busterSev: Severity, swapSev: Severity) => {
     const outputStrings = {
       tankSwap: Outputs.tankSwap,
       busterOnYou: Outputs.tankBusterOnYou,
@@ -179,31 +199,31 @@ export const Responses = {
     };
 
     // Note: busterSev and swapSev can be the same priority.
-    const tankSwapFunc = (data, matches, output) => {
+    const tankSwapFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (data.role === 'tank' && target !== data.me)
-        return output.tankSwap();
+        return output.tankSwap?.();
     };
-    const busterFunc = (data, matches, output) => {
+    const busterFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
 
       if (data.role === 'tank' && target !== data.me)
         return;
 
       if (target === data.me)
-        return output.busterOnYou();
-      return output.busterOnTarget({ player: data.ShortName(target) });
+        return output.busterOnYou?.();
+      return output.busterOnTarget?.({ player: data.ShortName(target) });
     };
 
     const combined = combineFuncs(defaultAlarmText(swapSev), tankSwapFunc,
         defaultAlertText(busterSev), busterFunc);
-    return (data, _, output) => {
+    return (data: Data, _: unknown, output: Output) => {
       // cactbot-builtin-response
       output.responseOutputStrings = outputStrings;
       return combined;
     };
   },
-  tankCleave: (sev) => (data, _, output) => {
+  tankCleave: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       cleaveOnYou: Outputs.tankCleaveOnYou,
@@ -211,231 +231,235 @@ export const Responses = {
       avoidCleave: Outputs.avoidTankCleave,
     };
     return {
-      [defaultInfoText(sev)]: (data, matches, output) => {
+      [defaultInfoText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const target = getTarget(matches);
         if (target === data.me)
-          return output.cleaveOnYou();
+          return output.cleaveOnYou?.();
         if (data.role === 'tank' || data.job === 'BLU') {
           // targetless tank cleave
           // BLU players should always get this generic cleave message.
           // We have no robust way to determine whether they have tank Mimicry on,
           // and it's really annoying for a BLU tank to be told to avoid cleaves when they can't.
-          return output.cleaveNoTarget();
+          return output.cleaveNoTarget?.();
         }
-        return output.avoidCleave();
+        return output.avoidCleave?.();
       },
     };
   },
-  miniBuster: (sev) => staticResponse(defaultInfoText(sev), Outputs.miniBuster),
-  aoe: (sev) => staticResponse(defaultInfoText(sev), Outputs.aoe),
-  bigAoe: (sev) => staticResponse(defaultInfoText(sev), Outputs.bigAoe),
-  spread: (sev) => staticResponse(defaultInfoText(sev), Outputs.spread),
+  miniBuster: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.miniBuster),
+  aoe: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.aoe),
+  bigAoe: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.bigAoe),
+  spread: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.spread),
   // for stack marker situations.
-  stackMarker: (sev) => staticResponse(defaultAlertText(sev), Outputs.stackMarker),
+  stackMarker: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.stackMarker),
   // for getting together without stack marker
-  getTogether: (sev) => staticResponse(defaultAlertText(sev), Outputs.getTogether),
-  stackMarkerOn: (sev) => (data, _, output) => {
+  getTogether: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.getTogether),
+  stackMarkerOn: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       stackOnYou: Outputs.stackOnYou,
       stackOnTarget: Outputs.stackOnPlayer,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const target = getTarget(matches);
         if (target === data.me)
-          return output.stackOnYou();
-        return output.stackOnTarget({ player: data.ShortName(target) });
+          return output.stackOnYou?.();
+        return output.stackOnTarget?.({ player: data.ShortName(target) });
       },
     };
   },
-  stackMiddle: (sev) => staticResponse(defaultInfoText(sev), Outputs.stackMiddle),
-  doritoStack: (sev) => staticResponse(defaultAlertText(sev), Outputs.doritoStack),
-  spreadThenStack: (sev) => staticResponse(defaultAlertText(sev), Outputs.spreadThenStack),
-  stackThenSpread: (sev) => staticResponse(defaultAlertText(sev), Outputs.stackThenSpread),
-  knockback: (sev) => staticResponse(defaultAlertText(sev), Outputs.knockback),
-  knockbackOn: (targetSev, otherSev) => {
+  stackMiddle: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.stackMiddle),
+  doritoStack: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.doritoStack),
+  spreadThenStack: (sev: Severity) => {
+    return staticResponse(defaultAlertText(sev), Outputs.spreadThenStack);
+  },
+  stackThenSpread: (sev: Severity) => {
+    return staticResponse(defaultAlertText(sev), Outputs.stackThenSpread);
+  },
+  knockback: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.knockback),
+  knockbackOn: (targetSev: Severity, otherSev: Severity) => {
     const outputStrings = {
       knockbackOnYou: Outputs.knockbackOnYou,
       knockbackOnTarget: Outputs.knockbackOnPlayer,
     };
 
-    const targetFunc = (data, matches, output) => {
+    const targetFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (target === data.me)
-        return output.knockbackOnYou();
+        return output.knockbackOnYou?.();
     };
 
-    const otherFunc = (data, matches, output) => {
+    const otherFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (target !== data.me)
-        return output.knockbackOnTarget({ player: data.ShortName(target) });
+        return output.knockbackOnTarget?.({ player: data.ShortName(target) });
     };
     const combined = combineFuncs(defaultInfoText(targetSev), targetFunc,
         defaultInfoText(otherSev), otherFunc);
-    return (data, _, output) => {
+    return (data: Data, _: unknown, output: Output) => {
       // cactbot-builtin-response
       output.responseOutputStrings = outputStrings;
       return combined;
     };
   },
-  lookTowards: (sev) => staticResponse(defaultInfoText(sev), Outputs.lookTowardsBoss),
-  lookAway: (sev) => staticResponse(defaultInfoText(sev), Outputs.lookAway),
-  lookAwayFromTarget: (sev) => (data, _, output) => {
+  lookTowards: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.lookTowardsBoss),
+  lookAway: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.lookAway),
+  lookAwayFromTarget: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       lookAwayFrom: Outputs.lookAwayFromTarget,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const target = getTarget(matches);
         if (target === data.me)
           return;
-        const name = isPlayerId(matches.targetId) ? data.ShortName(target) : target;
-        return output.lookAwayFrom({ name: name });
+        const name = isPlayerId(matches?.targetId) ? data.ShortName(target) : target;
+        return output.lookAwayFrom?.({ name: name });
       },
     };
   },
-  lookAwayFromSource: (sev) => (data, _, output) => {
+  lookAwayFromSource: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       lookAwayFrom: Outputs.lookAwayFromTarget,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const source = getSource(matches);
         if (source === data.me)
           return;
-        const name = isPlayerId(matches.sourceId) ? data.ShortName(source) : source;
-        return output.lookAwayFrom({ name: name });
+        const name = isPlayerId(matches?.sourceId) ? data.ShortName(source) : source;
+        return output.lookAwayFrom?.({ name: name });
       },
     };
   },
-  getBehind: (sev) => staticResponse(defaultAlertText(sev), Outputs.getBehind),
-  goFrontOrSides: (sev) => staticResponse(defaultAlertText(sev), Outputs.goFrontOrSides),
+  getBehind: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.getBehind),
+  goFrontOrSides: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.goFrontOrSides),
   // .getUnder() is used when you have to get into the bosses hitbox
-  getUnder: (sev) => staticResponse(defaultInfoText(sev), Outputs.getUnder),
+  getUnder: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.getUnder),
   // .getIn() is more like "get close but maybe even melee range is fine"
-  getIn: (sev) => staticResponse(defaultAlertText(sev), Outputs.in),
+  getIn: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.in),
   // .getOut() means get far away
-  getOut: (sev) => staticResponse(defaultAlertText(sev), Outputs.out),
-  outOfMelee: (sev) => staticResponse(defaultInfoText(sev), Outputs.outOfMelee),
-  getInThenOut: (sev) => staticResponse(defaultInfoText(sev), Outputs.inThenOut),
-  getOutThenIn: (sev) => staticResponse(defaultInfoText(sev), Outputs.outThenIn),
-  getBackThenFront: (sev) => staticResponse(defaultInfoText(sev), Outputs.backThenFront),
-  getFrontThenBack: (sev) => staticResponse(defaultInfoText(sev), Outputs.frontThenBack),
-  goMiddle: (sev) => staticResponse(defaultAlertText(sev), Outputs.goIntoMiddle),
-  goRight: (sev) => staticResponse(defaultAlertText(sev), Outputs.right),
-  goLeft: (sev) => staticResponse(defaultAlertText(sev), Outputs.left),
-  goWest: (sev) => staticResponse(defaultAlertText(sev), Outputs.getLeftAndWest),
-  goEast: (sev) => staticResponse(defaultAlertText(sev), Outputs.getRightAndEast),
-  goFrontBack: (sev) => staticResponse(defaultAlertText(sev), Outputs.goFrontBack),
-  goSides: (sev) => staticResponse(defaultAlertText(sev), Outputs.sides),
+  getOut: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.out),
+  outOfMelee: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.outOfMelee),
+  getInThenOut: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.inThenOut),
+  getOutThenIn: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.outThenIn),
+  getBackThenFront: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.backThenFront),
+  getFrontThenBack: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.frontThenBack),
+  goMiddle: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.goIntoMiddle),
+  goRight: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.right),
+  goLeft: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.left),
+  goWest: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.getLeftAndWest),
+  goEast: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.getRightAndEast),
+  goFrontBack: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.goFrontBack),
+  goSides: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.sides),
   // .killAdds() is used for adds that will always be available
-  killAdds: (sev) => staticResponse(defaultInfoText(sev), Outputs.killAdds),
+  killAdds: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.killAdds),
   // .killExtraAdd() is used for adds that appear if a mechanic was not played correctly
-  killExtraAdd: (sev) => staticResponse(defaultInfoText(sev), Outputs.killExtraAdd),
-  awayFromFront: (sev) => staticResponse(defaultAlertText(sev), Outputs.awayFromFront),
-  sleep: (sev) => (data, _, output) => {
+  killExtraAdd: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.killExtraAdd),
+  awayFromFront: (sev: Severity) => staticResponse(defaultAlertText(sev), Outputs.awayFromFront),
+  sleep: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       sleep: Outputs.sleepTarget,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const source = getSource(matches);
-        return output.sleep({ name: source });
+        return output.sleep?.({ name: source });
       },
     };
   },
-  stun: (sev) => (data, _, output) => {
+  stun: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       stun: Outputs.stunTarget,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const source = getSource(matches);
-        return output.stun({ name: source });
+        return output.stun?.({ name: source });
       },
     };
   },
-  interrupt: (sev) => (data, _, output) => {
+  interrupt: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       interrupt: Outputs.interruptTarget,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const source = getSource(matches);
-        return output.interrupt({ name: source });
+        return output.interrupt?.({ name: source });
       },
     };
   },
-  preyOn: (targetSev, otherSev) => {
+  preyOn: (targetSev: Severity, otherSev: Severity) => {
     const outputStrings = {
       preyOnYou: Outputs.preyOnYou,
       preyOnTarget: Outputs.preyOnPlayer,
     };
 
-    const targetFunc = (data, matches, output) => {
+    const targetFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (data.me === target)
-        return output.preyOnYou();
+        return output.preyOnYou?.();
     };
 
-    const otherFunc = (data, matches, output) => {
+    const otherFunc = (data: Data, matches: TargetedMatches, output: Output) => {
       const target = getTarget(matches);
       if (target !== data.me)
-        return output.preyOnTarget({ player: data.ShortName(target) });
+        return output.preyOnTarget?.({ player: data.ShortName(target) });
     };
 
     const combined = combineFuncs(defaultAlertText(targetSev), targetFunc,
         defaultInfoText(otherSev), otherFunc);
-    return (data, _, output) => {
+    return (data: Data, _: unknown, output: Output) => {
       // cactbot-builtin-response
       output.responseOutputStrings = outputStrings;
       return combined;
     };
   },
-  awayFrom: (sev) => (data, _, output) => {
+  awayFrom: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       awayFromGroup: Outputs.awayFromGroup,
       awayFromTarget: Outputs.awayFromPlayer,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const target = getTarget(matches);
         if (data.me === target)
-          return output.awayFromGroup();
-        return output.awayFromTarget({ player: data.ShortName(target) });
+          return output.awayFromGroup?.();
+        return output.awayFromTarget?.({ player: data.ShortName(target) });
       },
     };
   },
-  meteorOnYou: (sev) => staticResponse(defaultAlarmText(sev), Outputs.meteorOnYou),
-  stopMoving: (sev) => staticResponse(defaultAlarmText(sev), Outputs.stopMoving),
-  stopEverything: (sev) => staticResponse(defaultAlarmText(sev), Outputs.stopEverything),
+  meteorOnYou: (sev: Severity) => staticResponse(defaultAlarmText(sev), Outputs.meteorOnYou),
+  stopMoving: (sev: Severity) => staticResponse(defaultAlarmText(sev), Outputs.stopMoving),
+  stopEverything: (sev: Severity) => staticResponse(defaultAlarmText(sev), Outputs.stopEverything),
   // move away to dodge aoes
-  moveAway: (sev) => staticResponse(defaultInfoText(sev), Outputs.moveAway),
+  moveAway: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.moveAway),
   // move around (e.g. jumping) to avoid being frozen
-  moveAround: (sev) => staticResponse(defaultInfoText(sev), Outputs.moveAround),
-  breakChains: (sev) => staticResponse(defaultInfoText(sev), Outputs.breakChains),
-  moveChainsTogether: (sev) => staticResponse(defaultInfoText(sev),
+  moveAround: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.moveAround),
+  breakChains: (sev: Severity) => staticResponse(defaultInfoText(sev), Outputs.breakChains),
+  moveChainsTogether: (sev: Severity) => staticResponse(defaultInfoText(sev),
       Outputs.moveChainsTogether),
-  earthshaker: (sev) => (data, _, output) => {
+  earthshaker: (sev: Severity) => (data: Data, _: unknown, output: Output) => {
     // cactbot-builtin-response
     output.responseOutputStrings = {
       earthshaker: Outputs.earthshakerOnYou,
     };
     return {
-      [defaultAlertText(sev)]: (data, matches, output) => {
+      [defaultAlertText(sev)]: (data: Data, matches: TargetedMatches, output: Output) => {
         const target = getTarget(matches);
         if (target !== data.me)
           return;
-        return output.earthshaker();
+        return output.earthshaker?.();
       },
     };
   },
-  wakeUp: (sev) => staticResponse(defaultAlarmText(sev), Outputs.wakeUp),
+  wakeUp: (sev: Severity) => staticResponse(defaultAlarmText(sev), Outputs.wakeUp),
 };
```

File: resources/responses.ts [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+  };
+};
+
+type SingleSevToResponseFunc = (sev: Severity) => TargetedResponseFunc | StaticResponseFunc;
+type DoubleSevToResponseFunc = (targetSev: Severity, otherSev: Severity) => TargetedResponseFunc;
+type ResponsesMap = {
+  [response: string]: SingleSevToResponseFunc | DoubleSevToResponseFunc;
+};
+
+export const Responses: ResponsesMap = <const>{
```
Comment (panicstevenson) [inline resources/responses.ts:152]: This was mentioned offline, but should we enforce a styling here for `<...>` or `as ...`?

**Review by panicstevenson — APPROVED** [review]
LGTM. I wish there were a way to not have to have the optional chaining, but I think we can land this like this and maybe one day in the future we can have a 🎉 PR around removing the question marks from all the triggers.

File: resources/responses.ts (hunk 1 above)
Comment (quisquous) [reply resources/responses.ts]: Switched to `as const`.  I'll follow up with the eslint rule.

**Pull request merged**